### PR TITLE
Config templating

### PR DESCRIPTION
**Pull request by JorikSchellekens**
The synapse installer needs to template parts of the config. This adds some extra args to generate_config which is used to create the config. The extra args should be pretty usable in other situations. Some opinions on that would be nice.

I'd also like some advice on tying it into `load_or_generate_config` in `synapse/config/_base.py` without crowding the cli options.

### CODE REVIEW

File: synapse/config/_base.py [hunk 1]
```diff
@@ -181,6 +181,11 @@ def generate_config(
         generate_secrets=False,
         report_stats=None,
         open_private_ports=False,
+        listeners=None,
+        database=None,
+        tls_certificate_path=None,
+        tls_private_key_path=None,
+        acme_domain=None,
```
Comment (erikjohnston) [inline synapse/config/_base.py]: You should update the docstrings

File: synapse/config/database.py [hunk 2]
```diff
@@ -38,15 +38,17 @@ def read_config(self, config, **kwargs):
 
         self.set_databasepath(config.get("database_path"))
 
-    def generate_config_section(self, data_dir_path, **kwargs):
+    def generate_config_section(self, data_dir_path, database, **kwargs):
+        if not database:
+            database = "sqlite3"
```
Comment (erikjohnston) [inline synapse/config/database.py]: Given we're passing in a database path, not sure its useful to template out the database type?

File: synapse/config/tls.py [hunk 3]
```diff
@@ -239,12 +239,38 @@ def read_certificate_from_disk(self, require_cert_and_key):
...
     ):
+        """If the acme_domain is specified acme will be enabled.
+        If the TLS paths are not specified the default will be certs in the
+        config directory"""
+
         base_key_name = os.path.join(config_dir_path, server_name)
 
-        tls_certificate_path = base_key_name + ".tls.crt"
-        tls_private_key_path = base_key_name + ".tls.key"
+        if bool(tls_certificate_path) ^ bool(tls_private_key_path):
```
Comment (erikjohnston) [inline synapse/config/tls.py]:
```suggestion
        if bool(tls_certificate_path) != bool(tls_private_key_path):
```

I think is probably clearer tbh

File: synapse/config/tls.py [hunk 4]
```diff
@@ -343,7 +369,7 @@ def generate_config_section(
             # ACME support is disabled by default. Uncomment the following line
             # (and tls_certificate_path and tls_private_key_path above) to enable it.
             #
-            #enabled: true
+            %(acme_enabled)senabled: true
```
Comment (erikjohnston) [inline synapse/config/tls.py]: Can we do:

```suggestion
            enabled: %(acme_enabled)s
```

Instead? Then we can have the below always uncommented

File: synapse/config/server.py [hunk 5]
```diff
@@ -366,11 +369,70 @@ def generate_config_section(
...
+                        bind_addresses = listener.setdefault("bind_addresses", [])
+                        for address in private_addresses:
+                            if address not in bind_addresses:
+                                bind_addresses.append(address)
+
+                    unsecure_listeners.append(listener)
+
+            secure_http_bindings = indent(
+                yaml.dump(secure_listeners), " " * 10
+            ).lstrip()
```
Comment (erikjohnston) [inline synapse/config/server.py]: This indenting looks right, but also feels quite fragile. Can you add some unit tests that set various bits of config and makes sure the yaml round trips through the config generation correctly please?

File: synapse/config/database.py (hunk 2 above)
Comment (JorikSchellekens) [reply synapse/config/database.py]: We're not passing in a database path?

Comment (erikjohnston) [reply synapse/config/database.py]: Err no, but we are constructing one below

Comment (JorikSchellekens) [reply synapse/config/database.py]: I'm missing how we're inferring the database type from the path..

Comment (erikjohnston) [reply synapse/config/database.py]: If you're using postgres you don't use a path for the database, and you'll most likely have a bunch of other config

Comment (JorikSchellekens) [reply synapse/config/database.py]: Can I do something similar to what I've done with listeners: ie, just pass some dict and let the caller deal with it being valid and useful yaml?

File: synapse/config/server.py (hunk 5 above)
Comment (JorikSchellekens) [reply synapse/config/server.py]: Beauty is in delicate things. So I'll add some inelegant tests.

File: synapse/config/database.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/config/database.py]: Yup

**Pull request merged**